accepted/0000-hub_retail.md

```diff
@@ -0,0 +1,310 @@
+# Detailed design
+[design]: #detailed-design
+
+Images + Image metadata
+-----------------------
+
+Current situation: an image is built on Buildhost, then pushed to the Server's htdocs directory. Then the Server
+[creates a DB entry](https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/suse/manager/utils/SaltUtils.java#L1311)
+(including package list) and
+[a pillar file](https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/suse/manager/webui/services/SaltStateGeneratorService.java#L128).
+The pillar conains URL of image tarball, checksums and additional deployment details. It is used for image synchronization and deployment. These data are extracted
+on the build host at the end of build phase. Some of the data are only available on the build host and, not being contained in the image tarball, can't be extracted later.
+
+The SUMA API related to images is rather docker-centered. For Kiwi images it allows to list images and packages and delete an image.
+
+Use cases currently not covered:
+
+- Move image tarball to external HTTP server
+  In current implementation the image tarballs are stored in htdocs directory and distributed
+  via SUMA http server.
+  Branch Server via [Image Sync Formula](https://github.com/uyuni-project/retail/tree/master/image-sync-formula) downloads the tarball
+  from the URL in pillar, unpacks it and prepares the image for PXE service.
+
+  With Hub we can just extend this schema - tarball will be stored in Hub htdocs directory and copied to Peripheral servers via ISS.
+  This might not be optimal for certain cases (many small shops with combined Peripheral and Branch server).
+
+  So the alternative is to allow use of independent http servers and proxies (set up in customers infrastructure or in public cloud)
+  for distributing the tarballs. The http server structure does not have to strictly follow the Hub - Peripheral Server - Branch Server structure,
+  for example the Branch servers can download tarballs directly from datacenter or cloud.
+
+- Copy image metadata from one SUMA Server to another
+  Most common way of distributing image metadata is from Hub to Peripheral servers - this should be handled by
+  [ISS](https://github.com/uyuni-project/uyuni-rfc/blob/master/accepted/00063-hub-iss.md)
+
+  Another use case, not covered by the above, is this:
+  - Customer has one SUMA server in Lab environment and use it for developing images
+  - Only the images that pass all tests can be moved to production (which means move
+  them to Hub and distribute them)
+  - The approved image must stay the same (so it can't be rebuilt again)
+  - The development SUMA server should be as much separated from Hub as possible,
+  to be sure that the development does not break anything in production. Building
+  images on Hub is quite dangerous from this point of view.
+
+- Adjust deployment details
+  - currently the image pillars are write-only in SUMA, the users might want to change some details
```

Comment:
> currently the image pillars are write-only in SUMA

I don't follow. What do you mean by `write-only`? Do you mean read only?

Reply:
Java code just writes the pillars and forgets about them. There is no support for reading back and modifying.

Reply:
Ah, so by SUMA you meant just Java part. By SUMA I am always assuming the whole product including salt.